### PR TITLE
Fix/ Max order age on hanging orders

### PR DESCRIPTION
**Pull request by Nullably**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Refactor order_age fn to a utils file
- Max order age on active orders are recreated at refresh interval
- Max order age on hanging orders are first cancelled, then wait for cancellation event before recreate the hanging orders

### CODE REVIEW

**Comment by rapcmia** [comment]
I tried to test this with Binance (front-btc), I noticed that hanging orders are recreated on the set time for max_order_age however it does not work for active orders. 

First, I tested with order_refresh_time to `60s` and max_order_age to `30s` you will notice on the .GIF below that when it reaches 30s it recreates the orders for hanging orders but not for active orders. 
8fhULS10do


Second, since it's too low for max_order_age, I set up again with order_refresh_time of `60s`, max_order_age of `120s`, and order_refresh_tolerance of `1.0`. This way, we can retain the active orders until they reach `120s` and check if it recreates the order but it only recreates the hanging orders. See .GIF below:
dMkmjUMB8e

**Comment by nicolas-powertrade** [comment]
If hanging orders tracker module created in https://github.com/CoinAlpha/hummingbot/pull/3463 was to be merged, this could be implemented in pure market making strategy solving this issue

**Comment by keithbaum** [comment]
If hanging orders tracker module created in #3463 was to be merged, this could be implemented in pure market making strategy solving this issue

**Comment by Nullably** [comment]
> If hanging orders tracker module created in #3463 was to be merged, this could be implemented in pure market making strategy solving this issue

@keithbaum Nice! Yeah, I think that's a good idea.

**Comment by Nullably** [comment]
@rapcmia I fixed the issues, except for the case where max_order_age is below order_refresh_time, orders are canceled at max_order_age but will be created at the refresh time cycle. I think this is ok, to keep logic in the code simpler and for most cases  max_order_age shouldn't be below order_refresh_time anyway.

**Review by rapcmia — DISMISSED** [review]
Tested and confirmed working by QA team

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 1]
```diff
@@ -1136,6 +1141,49 @@ cdef class PureMarketMakingStrategy(StrategyBase):
...
+
+    cdef c_did_cancel_order(self, object cancelled_event):
+        cdef:
+            list orders = [o for o in self._to_recreate_hanging_orders if o.client_order_id == cancelled_event.order_id]
+        if orders:
+            self.c_recreate_hanging_order(orders[0])
+
+    @property
+    def to_recreate_hanging_orders(self) -> List[LimitOrder]:
+        return self._to_recreate_hanging_orders
```
Comment (CrimsonJacket) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx:1158]: Small matter, variable name suggest a return type of `bool`.
Might I suggest `_hanging_orders_to_recreate` instead

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 2]
```diff
@@ -681,13 +681,14 @@ cdef class PureMarketMakingStrategy(StrategyBase):
...
             self.c_cancel_hanging_orders()
             self.c_cancel_orders_below_min_spread()
-            refresh_proposal = self.c_aged_order_refresh()
-            # Firstly restore cancelled aged order
-            if refresh_proposal is not None:
-                self.c_execute_orders_proposal(refresh_proposal)
+            # refresh_proposal = self.c_aged_order_refresh()
+            # # Firstly restore cancelled aged order
+            # if refresh_proposal is not None:
+            #     self.c_execute_orders_proposal(refresh_proposal)
```
Comment (CrimsonJacket) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx:691]: I think this is a minor oversight. Let's remove it

**Review by CrimsonJacket — APPROVED** [review]
Some minor comments. But apart from that LGTM.

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx (hunk 1 above)
Comment (Nullably) [reply hummingbot/strategy/pure_market_making/pure_market_making.pyx:1158]: agree, good suggestion.

**Review by rxlxrxsx — APPROVED** [review]
Re-approving in behalf of QA team

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 3]
```diff
@@ -1083,9 +1076,20 @@ cdef class PureMarketMakingStrategy(StrategyBase):
...
-    # Cancel active non hanging orders
-    # Return value: whether order cancellation is deferred.
+    cdef c_cancel_active_orders_on_max_age_limit(self):
+        """
+        Cancels active non hanging orders if they are older than max age limit
+        """
+        cdef:
+            list active_orders = self.active_non_hanging_orders
+        if active_orders and any(order_age(o) > self._max_order_age for o in active_orders):
+            for order in active_orders:
```
Comment (aarmoa) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx:1086]: Is the intention to cancel all active orders? Shouldn't the logic cancel only the orders for which age is > _max_order_age?

**Review by aarmoa — APPROVED** [review]
I have discussed the comments with @Nullably. Changes look good to me.

**Pull request merged**